KAFKA-2825, KAFKA-2851: Extended existing ducktape replication tests to include controller failover

apovzner opened the pull request:
Plus fixed error copying key tab file when multiple builder jobs are running

apovzner committed:
Extended existing ducktape replication tests to include controller failover

apovzner committed:
Extended existing ducktape replication tests to include controller failover

apovzner committed:
Using random dir under /temp for local kdc files to avoid conflicts.

apovzner committed:
Using random dir under /temp for local kdc files to avoid conflicts.

apovzner committed:
Fixed usage of random dir under local /tmp for miniKdc files

apovzner committed:
Fixed usage of random dir under local /tmp for miniKdc files

apovzner committed:
Fix to make sure that temp dir for mini kcd files is removed after test finishes.

apovzner committed:
Fix to make sure that temp dir for mini kcd files is removed after test finishes.

benstopford commented:
I can't really comment on the Python but the logic looks good to me.

granders commented on tests/kafkatest/services/kafka/kafka.py:
```diff
@@ -274,24 +274,29 @@ def restart_node(self, node, clean_shutdown=True):
-    def leader(self, topic, partition=0):
-        """ Get the leader replica for the given topic and partition.
-        """
+    def query_zookeeper(self, zk_path):
```
This is handy - maybe it could go in `ZookeeperService` as a `query` method?

granders commented on tests/kafkatest/services/kafka/kafka.py:
```diff
@@ -303,6 +308,22 @@ def leader(self, topic, partition=0):
         return self.get_node(leader_idx)
 
+    def controller(self):
+        """ Get the controller ID
```
Outdated comment? I think this returns the controller node, not the controller ID

granders commented on tests/kafkatest/services/security/minikdc.py:
```diff
@@ -48,6 +49,16 @@ def start_node(self, node):
         self.logger.info("minikdc.properties")
         self.logger.info(props_file)
 
+        # create local temp dir where we will store keytab and krb5conf files
```
This local file stuff is all pretty much due to the fact that ducktape doesn't have a convenient mechanism to scp between worker nodes

I think it'll be a lot cleaner to add a utility method in `tests/kafkatest/utils/remoteaccount.py` to  encapsulate this behavior for now

Something like:

```
def scp(source_node, source_path, target_node, target_path):
  # source_node.account.copy_from(...) source_path -> local temp file
  # target_node.account.copy_to(...) local temp file -> target_path
```

Then you shouldn't have to track all of these "local" paths

apovzner replied:
@granders  Agreed that this looks cleaner. I will add that method and create a temp file with unique name (rather than the directory) in /tmp folder so that there is no conflict

granders commented:
@apovzner: A few minor comments, plus a possible approach to clean up `minikdc.py`, otherwise looks good.

apovzner committed:
Merge remote-tracking branch 'upstream/trunk' into cpkafka-86

apovzner committed:
KAFKA-2825 Moved query zookeeper method from KafkaService to ZookeeperService

apovzner committed:
KAFKA-2825 Moved query zookeeper method from KafkaService to ZookeeperService

Commit:
KAFKA-2746; Add support for using ConsumerGroupCommand on secure install

Author: Ashish Singh <asingh@cloudera.com>

Reviewers: Ismael Juma <ismael@juma.me.uk>, Jason Gustafson <jason@confluent.io>, Jun Rao <junrao@gmail.com>

Closes #534 from SinghAsDev/KAFKA-2746

apovzner committed:
KAFKA-2851 Added scp method to remote_account utils to scp between two remote nodes through unique local temp file

apovzner committed:
KAFKA-2851 Added scp method to remote_account utils to scp between two remote nodes through unique local temp file

apovzner committed:
KAFKA-2851: minor fix to format string in utils.remote_coount.scp method

apovzner committed:
KAFKA-2851: minor fix to format string in utils.remote_coount.scp method

Commit:
MINOR: add KStream merge operator

guozhangwang

Added KStreamBuilder.merge(KStream...).

Author: Yasuhiro Matsuda <yasuhiro@confluent.io>

Reviewers: Guozhang Wang

Closes #536 from ymatsuda/kstream_merge_operator

Commit:
MINOR: remove the group id from a restore consumer

guozhangwang
A restore consumer does not belong to a consumer group.

Author: Yasuhiro Matsuda <yasuhiro@confluent.io>

Reviewers: Guozhang Wang

Closes #543 from ymatsuda/no_group_for_restore_consumer

granders commented on tests/kafkatest/utils/remote_account.py:
```diff
@@ -29,4 +31,28 @@ def line_count(node, file):
+    :param target_node: destination node
+    :param target_path: path on the destination node to copy to
+    """
+    local_temp_dir = "/tmp"
```
A little confusing to set `local_temp_dir = "/tmp"` since it in theory should never be set to `/tmp`

I think this line is actually not necessary since `local_temp_dir` can be set inside the try block, and be available to the surrounding scope

Parth-Brahmbhatt committed:
KAFKA-2852; Updating the Authorizer CLI to use a consistent way to specify a list of values for a config options.

…ecify a list of values for a config options.

Author: Parth Brahmbhatt <brahmbhatt.parth@gmail.com>

Reviewers: Jun Rao <junrao@gmail.com>

Closes #545 from Parth-Brahmbhatt/KAFKA-2852

granders commented on tests/kafkatest/utils/remote_account.py:
```diff
@@ -29,4 +31,28 @@ def line_count(node, file):
+    except OSError as e:
+        raise Exception("Failed to create temporary local directory to scp %s to %s: %s" % (source_path, target_path, e.strerror))
+
+    local_temp_file = os.path.join(local_temp_dir, "tmpfile")
```
Maybe this copy block should be wrapped in `try/except/finally` so you can be sure to clean up temp files even if a copy phase fails. Then probably raise exception or rethrow if there was a problem copying

Also, for the cleanup in the finally block, consider
`shutil.rmtree(local_temp_dir, ignore_errors=True)`

guozhangwang committed:
MINOR: fix shutdownHook in ConsoleConsumer

Author: Confluent <confluent@Confluents-MacBook-Pro.local>

Reviewers: Jun Rao <junrao@gmail.com>

Closes #548 from guozhangwang/HFConsoleConsumer

ijuma committed:
MINOR: Introduce `producer.config` property to `ConsoleProducer`

This makes it easier to pass security properties in the same way
to `ConsoleConsumer` and `ConsoleProducer`.

Author: Ismael Juma <ismael@juma.me.uk>

Reviewers: Guozhang Wang <wangguoz@gmail.com>, Jun Rao <junrao@gmail.com>

Closes #544 from ijuma/producer-config-in-console-producer

hachikuji committed:
KAFKA-2841; safe group metadata cache loading/unloading

Author: Jason Gustafson <jason@confluent.io>

Reviewers: Guozhang Wang <wangguoz@gmail.com>, Jun Rao <junrao@gmail.com>

Closes #530 from hachikuji/KAFKA-2841

apovzner committed:
KAFKA-2851: clean up temp file even if scp fails

apovzner committed:
KAFKA-2851: clean up temp file even if scp fails

junrao committed:
trivial fix to 0.9.0 docs

granders commented on tests/kafkatest/utils/remote_account.py:
```diff
@@ -29,4 +32,25 @@ def line_count(node, file):
+    local_temp_file = os.path.join(local_temp_dir, "tmpfile")
+
+    try:
+        source_node.account.scp_from(source_path, local_temp_file)
```
Might want exception handling in case this scp fails for some reason?

apovzner committed:
KAFKA-2825 Using only PLAINTEXT and SASL_SSL security protocols for controller failover tests

apovzner committed:
KAFKA-2825 Using only PLAINTEXT and SASL_SSL security protocols for controller failover tests

granthenke committed:
KAFKA-2800; Update outdated dependencies

Changes:
* org.scala-lang:scala-library [2.10.5 -> 2.10.6]
   * Scala 2.10.6 resolves a license incompatibility in scala.util.Sorting
   * Otherwise identical to Scala 2.10.5
* org.xerial.snappy:snappy-java [1.1.1.7 -> 1.1.2]
   * Fixes SnappyOutputStream.close() is not idempotent
* net.jpountz.lz4:lz4 [1.2.0 -> 1.3]
* junit:junit [4.11 -> 4.12]
* org.easymock:easymock [3.3.1 -> 3.4]
* org.powermock:powermock-api-easymock [1.6.2 -> 1.6.3]
* org.powermock:powermock-module-junit4 [1.6.2 -> 1.6.3]
* org.slf4j:slf4j-api [1.7.6 -> 1.7.12]
* org.slf4j:slf4j-log4j12 [1.7.6 -> 1.7.12]
* com.fasterxml.jackson.jaxrs:jackson-jaxrs-json-provider [2.5.4 -> 2.6.3]
* com.fasterxml.jackson.core:jackson-databind [2.5.4 -> 2.6.3]
* org.eclipse.jetty:jetty-server [9.2.12.v20150709 -> 9.2.14.v20151106]
* org.eclipse.jetty:jetty-servlet [9.2.12.v20150709 -> 9.2.14.v20151106]
* org.bouncycastle:bcpkix-jdk15on [1.52 -> 1.53]
* net.sf.jopt-simple:jopt-simple [3.2 -> 4.9]
* removed explicit entry for org.objenesis:objenesis:2.2 (resolved transitively)

Author: Grant Henke <granthenke@gmail.com>

Reviewers: Ismael Juma <ismael@juma.me.uk>, Jun Rao <junrao@gmail.com>

Closes #513 from granthenke/update-deps

ijuma committed:
MINOR: Update to Gradle 2.9 and update generated `gradlew` file

More performance improvements:

"In many cases, Gradle 2.9 is much faster than Gradle 2.8 when performing incremental builds.

Very large builds (many thousands of source files) could see incremental build speeds up to 80% faster than 2.7 and up to 40% faster than 2.8.

Gradle now uses a more efficient mechanism to scan the filesystem, making up-to-date checks significantly faster. This improvement is only available when running Gradle with Java 7 or newer.

Other improvements have been made to speed-up include and exclude pattern evaluation; these improvements apply to all supported Java versions.

Gradle now uses much less memory than previous releases when performing incremental builds. By de-duplicating Strings used as file paths in internal caches, and by reducing the overhead of listing classes under test for Java projects, some builds use 30-70% less memory that Gradle 2.8."

https://docs.gradle.org/current/release-notes

Author: Ismael Juma <ismael@juma.me.uk>

Reviewers: Grant Henke, Guozhang Wang

Closes #549 from ijuma/gradle-2.9

ewencp committed:
KAFKA-2859: Fix deadlock in WorkerSourceTask.

Author: Ewen Cheslack-Postava <me@ewencp.org>

Reviewers: Gwen Shapira

Closes #554 from ewencp/kafka-2859-deadlock-worker-source-task

Commit:
KAFKA-2820: systest log level

Restores control over log level in system test service class KafkaService.

Author: Geoff Anderson <geoff@confluent.io>

Reviewers: Ismael Juma, Ewen Cheslack-Postava

Closes #538 from granders/KAFKA-2820-systest-log-level

Commit:
KAFKA-2845: new client old broker compatibility

Author: Geoff Anderson <geoff@confluent.io>

Reviewers: Ismael Juma, Guozhang Wang

Closes #537 from granders/KAFKA-2845-new-client-old-broker-compatibility

hachikuji committed:
KAFKA-2860: better handling of auto commit errors

Author: Jason Gustafson <jason@confluent.io>

Reviewers: Guozhang Wang

Closes #553 from hachikuji/KAFKA-2860

ijuma committed:
MINOR: Documentation improvements

* Fix typo in api.html
* Mark security features as beta quality (similar to new consumer). Is there better wording?
* Improve wording and clarify things in a number of places
* Improve layout of `pre` blocks (tested locally, which doesn't seem to use the same stylesheets as the deployed version)
* Use producer.config in console-producer.sh command
* Improve SASL documentation structure

Author: Ismael Juma <ismael@juma.me.uk>

Reviewers: Jun Rao, Magnus Edenhill, Gwen Shapira

Closes #550 from ijuma/documentation-improvements

benstopford committed:
KAFKA-2824: MiniKDC based tests don't run in VirtualBox

This is a hack which works. Is there a better way?

Build (v2) of the replication_test.py running here: http://jenkins.confluent.io/job/kafka_system_tests_branch_builder/185/

Author: Ben Stopford <benstopford@gmail.com>

Reviewers: Geoff Anderson, Gwen Shapira

Closes #520 from benstopford/fix-for-sasl-virtual-box

Commit:
KAFKA-2820: Remove log threshold on appender in tools-log4j.properties

Removed a config in tools-log4j.properties which prevented certain service classes from logging at TRACE level.

Author: Geoff Anderson <geoff@confluent.io>

Reviewers: Ewen Cheslack-Postava, Gwen Shapira

Closes #556 from granders/KAFKA-2820-systest-tool-loglevel

ewencp committed:
MINOR: Fix typo in sample Vagrantfile.local for AWS system tests

Author: Ewen Cheslack-Postava <me@ewencp.org>

Reviewers: Gwen Shapira

Closes #565 from ewencp/fix-aws-vagrantfile-local-example

ijuma committed:
KAFKA-2869; Host used by Authorizer should be IP address not hostname/IP

Author: Ismael Juma <ismael@juma.me.uk>

Reviewers: Jun Rao <junrao@gmail.com>

Closes #567 from ijuma/kafka-2869-host-used-by-authorizer-should-be-ip

ewencp committed:
KAFKA-2867: Fix missing WorkerSourceTask synchronization and handling of InterruptException.

Author: Ewen Cheslack-Postava <me@ewencp.org>

Reviewers: Jun Rao <junrao@gmail.com>

Closes #566 from ewencp/kafka-2867-fix-source-sync-and-interrupt

ijuma committed:
KAFKA-2863; Add a `close()` method to `Authorizer`

Author: Ismael Juma <ismael@juma.me.uk>

Reviewers: Jun Rao <junrao@gmail.com>

Closes #568 from ijuma/kafka-2863-authorizer-close

apovzner committed:
Merged miniKdc to run on VBox changes with KAFKA-2851 changes

apovzner committed:
Merge branch 'cpkafka-86' of https://github.com/apovzner/kafka into cpkafka-86

apovzner commented:
Closed this one and opened new PR https://github.com/apache/kafka/pull/570
